designcompose/src/main/java/com/android/designcompose/squoosh/SquooshRoot.kt

```diff
@@ -478,6 +417,110 @@ fun SquooshRoot(
+                        // While there are transitions to be run, we should run them; we just update the floats
+                        // in the mutable state. Those are then used by the render function, and we thus avoid
+                        // needing to recompose in order to propagate the animation state.
+                        //
+                        // We also complete transitions in this block, and that action does cause a recomposition
+                        // via the subscription that SquooshRoot makes to the InteractionState's list of transitions.
+                        while (interactionState.animations.isNotEmpty() || variantTransitions.transitions.isNotEmpty()) {
+                            // withFrameNanos runs its closure outside of this recomposition phase -- because the
+                            // body of the code only changes mutable state which is only used at render time,
+                            // Compose only re-runs the render block.
+                            withFrameNanos { frameTimeNanos ->
+                                val animState = HashMap(animationValues.value)
+                                for ((id, anim) in currentAnimations) {
+                                    // If we haven't started this animation yet, then start it now.
+                                    if (anim.startTimeNanos == 0L) {
+                                        anim.startTimeNanos = frameTimeNanos
+                                    }
+
+                                    val playTimeNanos = frameTimeNanos - anim.startTimeNanos
+
+                                    // Compute where it's meant to be, and update the value in animState.
+                                    val position =
+                                        anim.animation.getValueFromNanos(playTimeNanos)
+                                    animState[id] = position
+
+                                    // If the animation is complete, then we need to remove it from the transitions
+                                    // list, and apply it to the base interaction state.
+                                    if (anim.animation.isFinishedFromNanos(playTimeNanos)) {
+                                        animState.remove(id)
+                                        if (anim.action != null)
+                                            interactionState.squooshCompleteAnimatedAction(anim.action)
+                                        if (anim.variant != null)
+                                            variantTransitions.completedAnimatedVariant(anim.variant)
+                                    }
+                                }
+                                animationValues.value = animState
+                            }
+                        }
+                    }
+                }
+
                 val placeables = measurables.map { measurable ->
```

Comment:
This block, where we take `measurables` and generate `Constraints` for them from the output of layout should be upgraded. There are two types of `@Composable` child node for a `SquooshRoot`:

 1. Any node that uses interaction (or in the future that sets an accessibility property) causes a `@Composable` child to be generated. These synthetic `@Composable`s have no appearance, they're just there to capture events.
 2. Any `@Composable () -> Unit` customization results in a child `@Composable`. These are non-synthetic `@Composable` children.

The synthetic `@Composable`s can continue to be placed after layout has been evaluated. They just need to go exactly where the visual content is rendered from the `View` tree.

The non-synthetic `@Composable`s should probably be placed directly by Rust layout via a custom measure func. Rust Layout accepts a measure func for a node, and it passes some constraints and gets a `Size` back again. This is almost the same as what a Compose `Measurable` does -- it takes `Constraints` and gives you a `Placeable` with a width and height (the `measuredSize`). I think if we use Rust Layout's measure func to call `measurable.measure` for us, then we'll be able to support custom child composables that cause their DC Row/Column/Whatever to resize as they change size, or that can split text into multiple lines and occupy the correct height.

Reply:
Upon reflection, I think this approach can only work for a Layout that is the direct parent of the items that need to be custom measured. So it works for `SquooshRoot` in its current incarnation, but if we ever needed to change `SquooshRoot` so that it contained more of a tree (instead of squooshing everything into a single composable). I don't think it's possible to get measurables from deeper in the tree than immediate children.